### PR TITLE
added class exercises

### PR DESCRIPTION
**Pull request by AlejandroGispert**
Added Homework exercises
- Tic Tac Toe
- Voice reply App

Added in class exercises:
 -  characterFrequencies.js
 - creditCardNumber.js
 -  fibonacci.js   **corrected
 -  fizzFuzz.js
 - palindrome.js
 -  sentimentAnalyser.js

### CODE REVIEW

**Comment by nicoquiroga90** [comment]
Hello Alejandro! I see that you have completed many coding challenges, which is great practice for exercising logic and problem-solving skills. Well done! 💪🏼 

I will be reviewing the code and will try to provide some comments on your work. 🤓

File: javascript/javascript1/week4/characterFrequencies.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
+// Character frequencies
+// Write a function that counts the frequency of characters in a string:
+
+const getCharacterFrequencies = (str) => {
+  const characterArray = str.toLowerCase().split("").sort();
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/characterFrequencies.js:5]: Very well, making sure that the entire string is converted to lowercase before making comparisons. 🎖️

File: javascript/javascript1/week4/characterFrequencies.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+  let countOfCharacters = 0;
+  const characters = [];
+  let loopCounter = 0;
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/characterFrequencies.js:9]: 💡 We could remove this variable, It shouldn't be necessary for our function.

`let loopCounter = 0;`

File: javascript/javascript1/week4/characterFrequencies.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+  for (const char of characterArray) {
+    for (let i = 0; i < characterArray.length; i++) {
+      if (characterArray[i].match(char)) {
+        countOfCharacters = 0;
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/characterFrequencies.js:14]: ‼️ Here we are resetting the value to 0 and then adding one, and in the next iteration we assign the value to zero again.

File: javascript/javascript1/week4/characterFrequencies.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+const getCharacterFrequencies = (str) => {
+  const characterArray = str.toLowerCase().split("").sort();
+
+  let countOfCharacters = 0;
+  const characters = [];
+  let loopCounter = 0;
+
+  for (const char of characterArray) {
+    for (let i = 0; i < characterArray.length; i++) {
+      if (characterArray[i].match(char)) {
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/characterFrequencies.js]: The use of the match method is recommended for comparing a string with a regular expression. [Link](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/match)

💡 For comparisons between strings, we could use `===`

File: javascript/javascript1/week4/characterFrequencies.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+      if (loopCounter === 2) {
+        characters.push({ character: char, count: countOfCharacters });
+      }
+    }
+  }
+  const finalObject = { characters: characters, length: characterArray.length };
+  return finalObject;
+};
+console.log(getCharacterFrequencies("happyppp"));
+
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/characterFrequencies.js:36]: 🎖️ Great! The program produces results, perhaps we can simplify the logic:

💡 Alternatively, what if we approach the challenge as follows: What do we need to display?
- Total number of characters in the word (length)
- Number of times each character in the word repeats

The first one is simple because we have the method to count the string's length. We can obtain it at the end.

`length: characterArray.length`

The second one is more complex, but what we can do is create an object for each letter and increment +1 each time it repeats, using a loop. After counting, we can add the objects to an array.

`const characterCount = {};

  for (const char of characterArray) {
    characterCount[char] = (characterCount[char] || 0) + 1;
  }

  const characters = [];
  for (const char in characterCount) {
    characters.push({ character: char, count: characterCount[char] });
  }`

In conclusion, one way to approach the problem is to "break it into parts." It's not the only way, but it can serve as an approximation. 😊

File: javascript/javascript1/week4/creditCardInfo.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+  //or BIN, Bank Identification Number
+  //   const issuerIdentificationNumber = cardNumber.toString().slice(0, 6);
+  //   const lastDigit = cardNumber.toString().slice(-1);
+
+  const firstDigit = cardNumber.toString().slice(0, 1);
+  const firstTwoDigits = cardNumber.toString().slice(0, 2);
+
+  switch (true) {
+    case firstDigit == 4 &&
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/creditCardInfo.js]: ‼️ To make comparisons, it is recommended to use `===` for strict comparisons, avoiding implicit conversions.

⚠️ Comparison with `==` could yield undesired results.

The `==` operator compares two values after converting them to a common type (type coercion).
For example:

`5 == '5'  // true, because it converts '5' to a number before comparing

💡 With ===, you ensure that the comparison is accurate and it also provides more clarity to the code.


[More Info](https://www.freecodecamp.org/news/loose-vs-strict-equality-in-javascript/)

File: javascript/javascript1/week4/creditCardInfo.js [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  let cardType = "";
+
+  //or BIN, Bank Identification Number
+  //   const issuerIdentificationNumber = cardNumber.toString().slice(0, 6);
+  //   const lastDigit = cardNumber.toString().slice(-1);
+
+  const firstDigit = cardNumber.toString().slice(0, 1);
+  const firstTwoDigits = cardNumber.toString().slice(0, 2);
+
+  switch (true) {
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/creditCardInfo.js:14]: Nice use of Switch!  🤩

💡 Another alternative could be an if-else block.

File: javascript/javascript1/week4/creditCardInfo.js [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    case firstDigit == 4 &&
+      cardNumber.toString().length >= 13 &&
+      cardNumber.toString().length <= 16:
+      cardType = "Visa";
+
+      break;
+    case firstDigit == 5 && cardNumber.toString().length === 16:
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/creditCardInfo.js]: 🤓 To avoid repeating the use of toString, we could also convert the card number to a string beforehand and reuse it when necessary:

`const cardNumberStr = cardNumber.toString();`

File: javascript/javascript1/week4/creditCardNumber.js [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+console.log(formattedCreditCardObject);
+/*
+{
+  original: 123456789,
+  formatted: "1234 5678 9",
+}
+*/
+// Thins to consider:
+
+// What should happen if the function is called with an argument that is not a number?
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/creditCardNumber.js:57]: 🤝 Exactly! It is an important validation to avoid mishandling user input. 

My recommendation is to use a check like:

`if (typeof num !== 'number' || isNaN(num)) {
  return { error: 'Input must be a valid number' };
}
`

**typeof:** Verifies that the type of the value is number, which is a strict check that prevents empty strings, spaces, or any other non-numeric type from passing the validation.

**isNaN:** Ensures that the value is not NaN, covering cases like 0 / 0 or parseFloat('abc').

[About isNaN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/isNaN)

File: javascript/javascript1/week4/creditCardNumber.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+  for (const number of numArray) {
+    if (numArray.indexOf(number) < 4) {
+      result1 += number;
+    } else if (numArray.indexOf(number) >= 4 && numArray.indexOf(number) < 8) {
+      result2 += number;
+    } else if (numArray.indexOf(number) >= 8 && numArray.indexOf(number) < 12) {
+      result3 += number;
+    }
+  }
+  const finalResult = result1 + " " + result2 + " " + result3;
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/creditCardNumber.js]: Shouldn't there be 4 blocks of numbers to represent a credit card number, right? 😅

File: javascript/javascript1/week4/creditCardNumber.js [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+  return { original: num, formatted: finalResult };
+};
+
+const formattedCreditCardObject = formatCreditCardNumber(123456789);
+console.log(formattedCreditCardObject);
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/creditCardNumber.js:47]: Good logic, It fulfills the objective 🚀

A shorter path could also be to use a loop to iterate over the numbers and add a space every 4 places.

````
let formatted = "";

for (let i = 0; i < numStr.length; i++) {
  if (i > 0 && i % 4 === 0) {
    formatted += " ";
  }
  formatted += numStr[i];
}
````

File: javascript/javascript1/week4/fibonacci.js [hunk 12]
```diff
@@ -0,0 +1,24 @@
+//Fibonacci Sequence
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/fibonacci.js:1]: Well done!! 👏

File: javascript/javascript1/week4/fizzFuzz.js [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+// But for multiples of three print Fizz instead of the number and
+// for  the multiples of five print Buzz. For numbers which are multiples of
+// both three and five print FizzBuzz.
+
+// When that works. Make the two number for multiples into parameters.
+// So it can be called like this:
+
+const fizzBuzz = (num, num2) => {
+  let result = "";
+  for (let i = 0; i <= 100; i++) {
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/fizzFuzz.js]: Little detail: We want to start from 1, so in this case we can set our `i = 1`

File: javascript/javascript1/week4/fizzFuzz.js [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+    if (i % num !== 0) {
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/fizzFuzz.js]: ⚠️ There's a logic issue here. In this line, it's checking if the number i is not exactly divisible by num. If this condition is true, it means i **is not a multiple of num**. However, the logic of FizzBuzz requires printing "Fizz" for numbers that **are multiples of num**, not for those that aren't.

File: javascript/javascript1/week4/fizzFuzz.js [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+      result += "\n" + i;
+    } else if (i % num === 0) {
+      result += "\nFizz";
+    }
+    if (i % num === 0 && i % num2 === 0) {
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/fizzFuzz.js:15]: ⚠️ There's a problem in the code that causes the "Buzz" string to never be printed.

Currently, the check `if (i % num === 0 && i % num2 === 0)` is inside the for loop responsible for printing "Fizz". This means this check will only execute if i is divisible by num, preventing "Buzz" from being printed for numbers divisible only by num2.

To fix this, the `if (i % num2 === 0)` check should be outside the `else if` block so that it executes independently of whether i is divisible by num.

` if (i % num === 0 && i % num2 === 0) {
      result += "FizzBuzz\n";
    } else if (i % num === 0) {
      result += "Fizz\n";
    } else if (i % num2 === 0) {
      result += "Buzz\n";
    } else {
      result += i + "\n";`

File: javascript/javascript1/week4/highAndLow.js [hunk 16]
```diff
@@ -0,0 +1,10 @@
+function highAndLow(numbers) {
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/highAndLow.js:1]: Good job! 👏

File: javascript/javascript1/week4/palindrome.js [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+        substring.length > longest.length
+      ) {
+        longest = substring;
+      }
+    }
+  }
+
+  return longest;
+};
+console.log(longestPalindromic("lumuljhkhkh8"));
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/palindrome.js:25]: Awesome solution! 🚀

File: javascript/javascript1/week4/sentimentAnalyser.js [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+
+// Create a function that takes a string as a parameter.
+// calling the function will return an object with score,
+//  positiveWords and negativeWords.
+//  You decide how the score should be
+// implemented and what words are negative and positive.
+
+// Here is an example of using the function:
+
+const getSentimentScore = (str) => {
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/sentimentAnalyser.js:17]: 💡We could add at the beginning a conversion to lowercase to avoid problems with comparisons.

`str = str.toLowerCase();`

File: javascript/javascript1/week4/sentimentAnalyser.js [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+  const happyWordsArray = ["mega", "super", "awesome", "happy", "positive"];
+  const sadWordsArray = ["boring", "hate", "negative"];
+  const wordArray = str.split(" ");
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/sentimentAnalyser.js]: Well done!!  That way we make good use of includes! 🎖️

File: javascript/javascript1/week4/tic.js [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+
+  switch (true) {
+    case firstRow == "xxx" || firstRow == "ooo":
+      winner = firstRow[0];
+      break;
+    case secondRow == "xxx" || secondRow == "ooo":
+      winner = secondRow[0];
+      break;
+    case thirdRow == "xxx" || thirdRow == "ooo":
+      break;
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/tic.js:47]: It should be a winner case, right? 🔍

File: javascript/javascript1/week4/tic.js [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+    case antiDiagonal == "xxx" || antiDiagonal == "ooo":
+      winner = antiDiagonal[0];
+      break;
+    default:
+      winner = undefined;
+  }
+  let finalResult = {
+    winner: winner,
+    loser: winner === undefined ? undefined : winner === "x" ? "o" : "x",
+    hasEnded: winner != undefined ? true : false,
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/tic.js:69]: or

`hasEnded: !!winner`

File: javascript/javascript1/week4/tic.js [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+      break;
+    case antiDiagonal == "xxx" || antiDiagonal == "ooo":
+      winner = antiDiagonal[0];
+      break;
+    default:
+      winner = undefined;
+  }
+  let finalResult = {
+    winner: winner,
+    loser: winner === undefined ? undefined : winner === "x" ? "o" : "x",
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/tic.js:68]: Good use of the ternary operator! 🤩

It could also be:

`loser: winner === "x" ? "o" : winner === "o" ? "x" : undefined,`

File: javascript/javascript1/week4/voice.js [hunk 23]
```diff
@@ -0,0 +1,260 @@
+import "./styles.scss";
+
+// Or if installed from NPM to use with a bundler
+import Artyom from "artyom.js";
+// const artyom = require("artyom.js");
+const artyom = new Artyom();
+
+let catchedName = "";
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/voice.js]: Maybe a more descriptive variable name could be `capturedName`.

File: javascript/javascript1/week4/voice.js [hunk 24]
```diff
@@ -0,0 +1,260 @@
...
+const catchedToDo = [];
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/voice.js]: and `capturedToDo`.

File: javascript/javascript1/week4/voice.js [hunk 25]
```diff
@@ -0,0 +1,260 @@
...
+        if (
+          splittedArray[i] === "my" &&
+          splittedArray[i + 1] === "name" &&
+          splittedArray[i + 2] === "is"
+        ) {
+          catchedName = splittedArray[i + 3];
+          response = `Lovely to meet you ${catchedName}`;
+        }
+      }
+    } else if (catchedName != "") {
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/voice.js]: 💡 The strict comparison operator is recommend: 
    
`    } else if (catchedName !== "") { `

File: javascript/javascript1/week4/voice.js [hunk 26]
```diff
@@ -0,0 +1,260 @@
...
+      response = "You already told me your name " + catchedName;
+    }
+  }
+  //case 2
+  else if (command.toLowerCase().includes("what is my name")) {
+    if (catchedName == "") {
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/voice.js]: Also here 

`===`

File: javascript/javascript1/week4/voice.js [hunk 27]
```diff
@@ -0,0 +1,260 @@
...
+    },
+    onEnd: function () {
+      console.log("Dictation stopped by the user");
+    },
+  });
+
+  UserDictation.start();
+} else {
+  alert("no getReply function detected");
+}
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/voice.js:266]: Great job! Also, adding JS libraries to go one step further! 🚀🚀🚀

File: javascript/javascript1/week4/voice.js [hunk 28]
```diff
@@ -0,0 +1,260 @@
...
+      if (randomNumber % 2 === 0) {
+        button.innerHTML = "Talk now 😮";
+      } else {
+        button.innerHTML = "Talk now 🙂";
+      }
+    }, 100);
+    clearTimeout(timeoutId);
+
+    command = "";
+    timeoutId = setTimeout(() => {
```
Comment (nicoquiroga90) [inline javascript/javascript1/week4/voice.js:230]: Adding a `try...catch` block for error handling in asynchronous functions is definitely a good practice. 

It helps ensure that any errors occurring during asynchronous operations are properly caught and handled, preventing them from crashing the application or causing unexpected behavior. This can significantly improve the robustness and reliability of the code.

**Review by nicoquiroga90 — CHANGES_REQUESTED** [review]
You've passed challenging code tests, demonstrating a great ability to solve problems and apply them to code! Congratulations! 👏🏼👏🏼👏🏼

And remember:

_"True mastery is being an eternal apprentice."_  🥋

File: javascript/javascript1/week4/characterFrequencies.js (hunk 4 above)
Comment (AlejandroGispert) [reply javascript/javascript1/week4/characterFrequencies.js]: i fixed it now

File: javascript/javascript1/week4/characterFrequencies.js (hunk 5 above)
Comment (AlejandroGispert) [reply javascript/javascript1/week4/characterFrequencies.js:36]: Wow, your solution is so much cleaner! !
i am trying to make sense of this line  
characterCount[char] = (characterCount[char] || 0) + 1;
how is it possible to use the OR || like this?

File: javascript/javascript1/week4/characterFrequencies.js (hunk 2 above)
Comment (AlejandroGispert) [reply javascript/javascript1/week4/characterFrequencies.js:9]: further down, in line 26 its necesary to count tle loops i am in, if not i will end up with 

characters: [
    { character: 'a', count: 1 },
    { character: 'h', count: 1 },
    { character: 'p', count: 5 },
    { character: 'p', count: 5 },
    { character: 'p', count: 5 },
    { character: 'p', count: 5 },
    { character: 'p', count: 5 },
    { character: 'y', count: 1 }
  ],
  length: 8

File: javascript/javascript1/week4/characterFrequencies.js (hunk 3 above)
Comment (AlejandroGispert) [reply javascript/javascript1/week4/characterFrequencies.js:14]: yes, i reset it, everytime it changes character. I realize the code is not so easy to read. I prefer your solution!

File: javascript/javascript1/week4/creditCardInfo.js (hunk 6 above)
Comment (AlejandroGispert) [reply javascript/javascript1/week4/creditCardInfo.js]: Done!

File: javascript/javascript1/week4/creditCardInfo.js (hunk 8 above)
Comment (AlejandroGispert) [reply javascript/javascript1/week4/creditCardInfo.js]: done!

File: javascript/javascript1/week4/creditCardNumber.js (hunk 9 above)
Comment (AlejandroGispert) [reply javascript/javascript1/week4/creditCardNumber.js:57]: Awesome information!! i will note down this one!

File: javascript/javascript1/week4/creditCardNumber.js (hunk 10 above)
Comment (AlejandroGispert) [reply javascript/javascript1/week4/creditCardNumber.js]: done!

File: javascript/javascript1/week4/creditCardNumber.js (hunk 11 above)
Comment (AlejandroGispert) [reply javascript/javascript1/week4/creditCardNumber.js:47]: THIS IS A GREAT SOLUTION! noted!

File: javascript/javascript1/week4/fizzFuzz.js (hunk 14 above)
Comment (AlejandroGispert) [reply javascript/javascript1/week4/fizzFuzz.js]: on line 16, it shows that is printing a new line only, and on the line 17 it checks for the exact divisibles for num

File: javascript/javascript1/week4/sentimentAnalyser.js (hunk 18 above)
Comment (AlejandroGispert) [reply javascript/javascript1/week4/sentimentAnalyser.js:17]: right! done

File: javascript/javascript1/week4/tic.js (hunk 20 above)
Comment (AlejandroGispert) [reply javascript/javascript1/week4/tic.js:47]: 😔 right

File: javascript/javascript1/week4/tic.js (hunk 21 above)
Comment (AlejandroGispert) [reply javascript/javascript1/week4/tic.js:69]: what does this mean? i didnt get it !!winner.    ?

File: javascript/javascript1/week4/voice.js (hunk 28 above)
Comment (AlejandroGispert) [reply javascript/javascript1/week4/voice.js:230]: i did it, but try ... catch is still new ground for me
